### PR TITLE
Azure RedHat OpenShift 2022-09-04

### PR DESCRIPTION
**Pull request by ross-bryan**
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  September 7th
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  September 20th
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    - [x] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by raosuhas** [comment]
Pleae follow the instructions in the checklist before sumitting the PR. This PR is not created correctly : 

image

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (raosuhas) [inline specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:57]: >get [](http://example.com/codeflow?start=7&length=3)

was this meant to be a POST ? THis current path will be interpreted by arm as a collection get for the resource. If you keep it as a GET the InstallVersions object would need to be a ARM resource.

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (raosuhas) [inline specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:101]: shoud this be "Lists...for the openshiftcluster" ? This would not list everything in the subscription right ?

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (raosuhas) [inline specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:538]: >childResourceName [](http://example.com/codeflow?start=150&length=17)

can you please update with descriptive variables .. i.e syncsetname vs childresourcename

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (raosuhas) [inline specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:1018]: This needs to be defined as a ProxyResource, use the definition from commontypes to define it as such

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (raosuhas) [inline specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:1048]: free form objects are not allowed. Please define the schema for this

**Comment by ross-bryan** [comment]
@raosuhas thank you for the feedback. I was still working on some of the definitions/descriptions. I wasn't aware that the PR was adding/removing the labels for you to review it so soon. I will address all your comments in my next push

**Comment by raosuhas** [comment]
> @raosuhas thank you for the feedback. I was still working on some of the definitions/descriptions. I wasn't aware that the PR was adding/removing the labels for you to review it so soon. I will address all your comments in my next push

np .. there seems to be a bug in the label setting alogrithm which the team is working on , so I am just manually checking PRs for now .. so I might be looking at some PRs incorrectly

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (ross-bryan) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json]: yes, correct. updated

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (ross-bryan) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json]: yes, updated

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (ross-bryan) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:1027]: I will do that. It should be looking better now

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (ross-bryan) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:57]: we want it to be a GET request - the idea is for the ARO service to provide same command as AKS, they have `az aks get-versions` and we are working on implementing the `az aro get-versions` equivalent.

I am trying to see more of how AKS achieved this, it looks like the endpoint for their command is as follows:
```
'https://management.azure.com/subscriptions/fe16a035-e540-4ab7-80d9-373fa9a3d6ae/providers/Microsoft.ContainerService/locations/eastus/orchestrators?api-version=2019-04-01&resource-type=managedClusters'
```
is the orchestrators an ARM resource? I see this for AKS service https://github.com/Azure/azure-rest-api-specs/blob/3ec782ff73d2c35f23978d2c6738cc9c85016531/specification/containerservice/resource-manager/Microsoft.ContainerService/stable/2022-06-01/managedClusters.json#L2699-L2703

**Comment by raosuhas** [comment]
Fixed the labels , please remove ARMCHangesRequested label when you want the ARM reviewer to look at this again.

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (ross-bryan) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:57]: @raosuhas can you review my response here?

Comment (raosuhas) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:57]: we are trying to discourage RPs from using this model 

For your scenario you can achieve pretty much the same thing by modelling this as a GET on a collection of "installVersions" i.e 

GET  ...crosoft.RedHatOpenShift/locations/{location}/installversions" 

THis returns a paginated response like so : [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#pagination](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#pagination)

and each entry of the value array is a Proxy resource of the "InstallVersion" type. The only difference from the AKS model is that each installVersion is a Proxy ARM  type.

Comment (ross-bryan) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:57]: @raosuhas I have refactored my code and api /swagger definition such that this is a `GET` request which returns a `List` of `OpenShiftVersion`'s  that are available.

Can you please review? I do not seem to be able to edit the labels on the PR

**Comment by ross-bryan** [comment]
@leni-msft would you be able to review this PR? I see you have worked with ARO team in our previous API release.

**Comment by leni-msft** [comment]
@ross-bryan Adding new API requires ARM sign-off. I see ARM has requested changes, please check.

**Comment by ross-bryan** [comment]
Hi @leni-msft , I made the changes that were requested see this thread https://github.com/Azure/azure-rest-api-specs/pull/20463#discussion_r975526618

Also, this was passing all validation last week. Now there are new rules that are added that is causing more validation to fail. It is unreasonable for me to keep trying to pass linting validation when it is a constantly moving target. Can you please advise? 

It seems the label bot is broken in this PR.

File: specification/redhatopenshift/resource-manager/readme.azureresourceschema.md
Comment (leni-msft) [inline specification/redhatopenshift/resource-manager/readme.azureresourceschema.md:1]: this file is not needed, pls delete

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (leni-msft) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:57]: @ross-bryan I think @raosuhas is also suggesting changing the path to `/installversions`, so in this case the resource type is "InstallVersion"

**Review by leniatgh — CHANGES_REQUESTED** [review]

File: specification/redhatopenshift/resource-manager/readme.azureresourceschema.md
Comment (ventifus) [reply specification/redhatopenshift/resource-manager/readme.azureresourceschema.md]: deleted.

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (ventifus) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:57]: @raosuhas @leni-msft We have changed this resource path to 

GET ...crosoft.RedHatOpenShift/locations/{location}/openshiftversions"

**Comment by ventifus** [comment]
Hi @leni-msft I am working on making the checks pass for this. I have some questions:

The checks were indicating that we must have a `pattern` parameter for `resourceName` and `childResourceName`. We don't have any name constraints on our side, but when I add a pattern now I'm getting a violation "1036 - ConstraintChanged". Please let me know if there's some way of working around this conflict.

I don't understand why we're getting CreateOperationAsyncResponseValidation errors for several of our PUT operations. These are not async long-running operations, so we are not setting x-ms-long-running-operation. Can you clarify what the expectation is here?

**Comment by leni-msft** [comment]
> Hi @leni-msft I am working on making the checks pass for this. I have some questions:
> 
> The checks were indicating that we must have a `pattern` parameter for `resourceName` and `childResourceName`. We don't have any name constraints on our side, but when I add a pattern now I'm getting a violation "1036 - ConstraintChanged". Please let me know if there's some way of working around this conflict.
> 
> I don't understand why we're getting CreateOperationAsyncResponseValidation errors for several of our PUT operations. These are not async long-running operations, so we are not setting x-ms-long-running-operation. Can you clarify what the expectation is here?

Adding pattern in resource name is a new ARM rule, please request breaking change review for "1036 - ConstraintChanged", 

For CreateOperationAsyncResponseValidation, since @jianyexi is OOF, @rkmanda would you be able to take a look if it's a valid issue?

**Comment by leni-msft** [comment]
@ross-bryan please avoid force pushing the branch - it's hard to review what's the new change

**Comment by ventifus** [comment]
Thank you for the feedback. I found the code that implements CreateOperationAsyncResponseValidation, and it looks like _any_ PUT that results in a 201 Created is considered a long running operation. To work around this, we are removing the 201 responses and will use 200 OK instead.

**Comment by leni-msft** [comment]
> Thank you for the feedback. I found the code that implements CreateOperationAsyncResponseValidation, and it looks like _any_ PUT that results in a 201 Created is considered a long running operation. To work around this, we are removing the 201 responses and will use 200 OK instead.

@ventifus Does your service return 201? if so, you must define 201 in swagger. To workaround the lint error, we can suppress it. See https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process

> _any_ PUT that results in a 201 Created is considered a long running operation

+@jianyexi to check the logic

**Comment by ventifus** [comment]
HI @leni-msft Going back to the "pattern" restrictions and the resulting breaking change, I joined the Breaking Change office hours and they advised that the requirement for pattern restrictions is just a "staging" check and that we should not implement them at this time. (Also they noted that the client does not perform validation and so the restrictions would not have any effect anyway.) I'll submit a new commit reverting the pattern restrictions.

Regarding the 201 responses, yes we had originally designed our API to return 201 Created for PUT requests as that made the most sense. We would prefer to use 201, I will look into the suppression process.

**Comment by ventifus** [comment]
@leni-msft I have removed the pattern restrictions and restored the 201 Created responses for our PUTs. Since both of these are staging checks, do we still need to go through the swagger suppression process? The regular "Swagger LintDiff" is reporting green.

**Comment by leni-msft** [comment]
> @leni-msft I have removed the pattern restrictions and restored the 201 Created responses for our PUTs. Since both of these are staging checks, do we still need to go through the swagger suppression process? The regular "Swagger LintDiff" is reporting green.

You're right, no need for the staging lint check.

**Comment by ross-bryan** [comment]
@jianyexi @leni-msft can we get an approval here? This PR has been open for many many weeks now. We are delayed on delivering highly demanded features such as multi-version support because of this PR pending approval. Please either request changes, or grant us your blessings and approval :pray:

**Comment by leni-msft** [comment]
> If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them.

Hi @ross-bryan , this PR is blocked on ARM review. I will merge this PR after ARM sign-off.

**Review by leniatgh — APPROVED** [review]

**Comment by ventifus** [comment]
Looks like this is not getting an ARM review because it still has the tag ARMChangesRequested. I don't have permissions to remove it, @leni-msft @raosuhas can you help?

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (rkmanda) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:2335]: I dont see that this comment was addressed

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (rkmanda) [inline specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:2347]: >systemData [](http://example.com/codeflow?start=9&length=10)

SyncSetUpdate is defined as a parameter and it doesnt make sense for paramaters to have readOnly properties

Comment (ventifus) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:2347]: @rkmanda It is our intention that this systemData property is the Azure system metadata. In that context, should the systemData property be omitted from the swagger entirely?
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (ventifus) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:2335]: @rkmanda @raosuhas can you expand on what you want here? What I'm seeing is that SyncSetProperties has a single "resources" string property. In the implementation, "resources" will be a base64-encoded blob.

File: specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json
Comment (ventifus) [reply specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2022-09-04/redhatopenshift.json:2347]: I looked in the repo for other users of systemData, and our definitions seem to match theirs. Compare with adp.json [DataPoolPatch](https://github.com/Azure/azure-rest-api-specs/blob/4ac371de344db713847ff4d2ac1eae1e2223179a/specification/adp/resource-manager/Microsoft.AutonomousDevelopmentPlatform/preview/2021-11-01-preview/adp.json#L828) and used in a [PATCH method here](https://github.com/Azure/azure-rest-api-specs/blob/4ac371de344db713847ff4d2ac1eae1e2223179a/specification/adp/resource-manager/Microsoft.AutonomousDevelopmentPlatform/preview/2021-11-01-preview/adp.json#L515).

**Comment by ventifus** [comment]
Please close this PR in favor of #21248.

**Pull request closed without merging**
